## **Assignment on Data Preprocessing**

In [2]:
import numpy as np
import pandas as pd

In [3]:
from google.colab import files
uploaded=files.upload()

Saving titanic.csv to titanic.csv


In [5]:
# Reading dataset
df = pd.read_csv('titanic.csv')

In [6]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [7]:
#finding column values
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [8]:
#checking unique values
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [9]:
#dataset shape
df.shape

(891, 12)

In [10]:
#checking datatypes
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [11]:
#checking null values
df.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [13]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Null values are present in Age, Cabin and Embarked fields

Out of 891 rows, 687 missing values are seen in Cabin field. It would be difficult to handle the null values in field, as it may show variation in the model output, if mode value is used. So, this can be dropped.

In [18]:
df1=df.drop(['Cabin'],axis=1)
df1

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[891 rows x 11 columns]

A constant value 0 cannot be added to Age field, so we are not using simple imouter function to handle Age value

In [20]:
df1['Age']=df1['Age'].fillna(df1['Age'].median())

In [21]:
df1['Embarked']=df1['Embarked'].fillna(df1['Embarked'].mode().iloc[0])
df1.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Thus, the missing values are handled


In [22]:
df1.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object

As we need to male the feature values under int/float datatype, we need to do encoding to categorical fields.

We have 3 object fields. As the Name field will not have much significance in the model, this can be dropped.
The other fields are Sex and Embarked. Now, we can check for the number of unique values in each field

In [23]:
df1.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Embarked         3
dtype: int64

In [24]:
#Label encoding for the fields "Sex" and "Embarked"
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()
df1['Sex']= label_encoder.fit_transform(df1['Sex'])
df1['Sex'].unique()

array([1, 0])

In [25]:
df1['Embarked']= label_encoder.fit_transform(df1['Embarked'])
df1['Embarked'].unique()

array([2, 0, 1])

In [26]:
df1

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name  Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    1  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1   
2                               Heikkinen, Miss. Laina    0  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1   
4                             Allen, Mr. William Henry    1  35.0      0   
..                                                 ...  ...   ...    ...   
886                              Montvila, Rev. Juozas    1  27.0      0   
887                       Graham, Miss. Margaret Edith    0  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"    0  28.0      1   
889                              Behr, Mr. Karl Howell    1  26.0      0   
890                                Dooley, Mr. Patrick    1  32.0      0   

     Parch            Ticket     Fare  Embarked  
0        0         A/5 21171   7.2500         2  
1        0          PC 17599  71.2833         0  
2        0  STON/O2. 3101282   7.9250         2  
3        0            113803  53.1000         2  
4        0            373450   8.0500         2  
..     ...               ...      ...       ...  
886      0            211536  13.0000         2  
887      0            112053  30.0000         2  
888      2        W./C. 6607  23.4500         2  
889      0            111369  30.0000         0  
890      0            370376   7.7500         1  

[891 rows x 11 columns]

In [27]:
#Dropping Name field
df2=df1.drop(['Name'],axis=1)
df2

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch            Ticket  \
0              1         0       3    1  22.0      1      0         A/5 21171   
1              2         1       1    0  38.0      1      0          PC 17599   
2              3         1       3    0  26.0      0      0  STON/O2. 3101282   
3              4         1       1    0  35.0      1      0            113803   
4              5         0       3    1  35.0      0      0            373450   
..           ...       ...     ...  ...   ...    ...    ...               ...   
886          887         0       2    1  27.0      0      0            211536   
887          888         1       1    0  19.0      0      0            112053   
888          889         0       3    0  28.0      1      2        W./C. 6607   
889          890         1       1    1  26.0      0      0            111369   
890          891         0       3    1  32.0      0      0            370376   

        Fare  Embarked  
0     7.2500         2  
1    71.2833         0  
2     7.9250         2  
3    53.1000         2  
4     8.0500         2  
..       ...       ...  
886  13.0000         2  
887  30.0000         2  
888  23.4500         2  
889  30.0000         0  
890   7.7500         1  

[891 rows x 10 columns]

In [28]:
df2.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Sex              int64
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked         int64
dtype: object

Thus, all the fields got changed to numerical fields.

In [29]:
# Splitting the data to Train and Test class
y=df2['Survived']
X=df2.drop(['Survived'],axis=1)
X.describe()

PassengerId      Pclass         Sex         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    2.308642    0.647587   29.361582    0.523008   
std     257.353842    0.836071    0.477990   13.019697    1.102743   
min       1.000000    1.000000    0.000000    0.420000    0.000000   
25%     223.500000    2.000000    0.000000   22.000000    0.000000   
50%     446.000000    3.000000    1.000000   28.000000    0.000000   
75%     668.500000    3.000000    1.000000   35.000000    1.000000   
max     891.000000    3.000000    1.000000   80.000000    8.000000   

            Parch        Fare    Embarked  
count  891.000000  891.000000  891.000000  
mean     0.381594   32.204208    1.536476  
std      0.806057   49.693429    0.791503  
min      0.000000    0.000000    0.000000  
25%      0.000000    7.910400    1.000000  
50%      0.000000   14.454200    2.000000  
75%      0.000000   31.000000    2.000000  
max      6.000000  512.329200    2.000000

In [31]:
X=df2.drop(['Ticket'],axis=1)
X.describe()

PassengerId    Survived      Pclass         Sex         Age  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642    0.647587   29.361582   
std     257.353842    0.486592    0.836071    0.477990   13.019697   
min       1.000000    0.000000    1.000000    0.000000    0.420000   
25%     223.500000    0.000000    2.000000    0.000000   22.000000   
50%     446.000000    0.000000    3.000000    1.000000   28.000000   
75%     668.500000    1.000000    3.000000    1.000000   35.000000   
max     891.000000    1.000000    3.000000    1.000000   80.000000   

            SibSp       Parch        Fare    Embarked  
count  891.000000  891.000000  891.000000  891.000000  
mean     0.523008    0.381594   32.204208    1.536476  
std      1.102743    0.806057   49.693429    0.791503  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    7.910400    1.000000  
50%      0.000000    0.000000   14.454200    2.000000  
75%      1.000000    0.000000   31.000000    2.000000  
max      8.000000    6.000000  512.329200    2.000000

**Feature Scaling:**

Using this technique, we can standardize the independent features present in the data in a fixed range

In [32]:

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [33]:
#Min Max scaler
min_max=preprocessing.MinMaxScaler(feature_range=(0,1))
X=min_max.fit_transform(X)
X=pd.DataFrame(X)
X.describe()

0           1           2           3           4           5  \
count  891.000000  891.000000  891.000000  891.000000  891.000000  891.000000   
mean     0.500000    0.383838    0.654321    0.647587    0.363679    0.065376   
std      0.289162    0.486592    0.418036    0.477990    0.163605    0.137843   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.250000    0.000000    0.500000    0.000000    0.271174    0.000000   
50%      0.500000    0.000000    1.000000    1.000000    0.346569    0.000000   
75%      0.750000    1.000000    1.000000    1.000000    0.434531    0.125000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

                6           7           8  
count  891.000000  891.000000  891.000000  
mean     0.063599    0.062858    0.768238  
std      0.134343    0.096995    0.395752  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.015440    0.500000  
50%      0.000000    0.028213    1.000000  
75%      0.000000    0.060508    1.000000  
max      1.000000    1.000000    1.000000

From the above, it is seen that the values are standardized between a min value of 0 and max value of 1

In [35]:
#standard scaler
standardisation=preprocessing.StandardScaler()
X=standardisation.fit_transform(X)
X=pd.DataFrame(X)
X.describe()

0             1             2             3             4  \
count  8.910000e+02  8.910000e+02  8.910000e+02  8.910000e+02  8.910000e+02   
mean   6.379733e-17  3.987333e-17 -8.772133e-17 -1.156327e-16  1.913920e-16   
std    1.000562e+00  1.000562e+00  1.000562e+00  1.000562e+00  1.000562e+00   
min   -1.730108e+00 -7.892723e-01 -1.566107e+00 -1.355574e+00 -2.224156e+00   
25%   -8.650540e-01 -7.892723e-01 -3.693648e-01 -1.355574e+00 -5.657365e-01   
50%    0.000000e+00 -7.892723e-01  8.273772e-01  7.376951e-01 -1.046374e-01   
75%    8.650540e-01  1.266990e+00  8.273772e-01  7.376951e-01  4.333115e-01   
max    1.730108e+00  1.266990e+00  8.273772e-01  7.376951e-01  3.891554e+00   

                  5             6             7             8  
count  8.910000e+02  8.910000e+02  8.910000e+02  8.910000e+02  
mean   4.386066e-17  7.376566e-17  1.196200e-17 -4.984166e-17  
std    1.000562e+00  1.000562e+00  1.000562e+00  1.000562e+00  
min   -4.745452e-01 -4.736736e-01 -6.484217e-01 -1.942303e+00  
25%   -4.745452e-01 -4.736736e-01 -4.891482e-01 -6.781745e-01  
50%   -4.745452e-01 -4.736736e-01 -3.573909e-01  5.859541e-01  
75%    4.327934e-01 -4.736736e-01 -2.424635e-02  5.859541e-01  
max    6.784163e+00  6.974147e+00  9.667167e+00  5.859541e-01